File: servers/elixir/lib/haveapi/validator/length.ex
defmodule HaveAPI.Validator.Length do
  use HaveAPI.Validator

  def init(opts) do
    if opts[:equals] && (opts[:min] || opts[:max]) do
      raise "cannot mix equals and min/max"
    end

    mkopts(opts, opts[:equals], opts[:min], opts[:max])
  end

  def validate(opts, v, _params), do: do_validate(opts, String.length(v))

  defp mkopts(opts, equals, nil, nil) when is_integer(equals) do
    %{
      equals: equals,
      message: opts[:message] || "length has to be #{equals}"
    }
  end

  defp mkopts(opts, nil, min, nil) when is_integer(min) do
    %{
      min: min,
      message: opts[:message] || "length has to be minimally #{min}"
    }
  end

  defp mkopts(opts, nil, nil, max) when is_integer(max) do
    %{
      max: max,
      message: opts[:message] || "length has to be maximally #{max}"
    }
  end

  defp mkopts(opts, nil, min, max) when is_integer(min) and is_integer(max) do
    %{
      min: min,
      max: max,
      message: opts[:message] || "length has to be in range <#{min}, #{max}>"
    }
  end

  defp do_validate(%{equals: n} = opts, len) when is_integer(n) do
    return(opts, n == len)
  end

  defp do_validate(%{min: min, max: max} = opts, len) when is_integer(min) and is_integer(max) do
    return(opts, len >= min && len <= max)
  end

  defp do_validate(%{min: min} = opts, len) when is_integer(min) do
    return(opts, len >= min)
  end

  defp do_validate(%{max: max} = opts, len) when is_integer(max) do
    return(opts, len <= max)
  end

  defp return(_opts, true), do: :ok
  defp return(opts, false), do: {:error, [opts.message]}
end

File: servers/ruby/lib/haveapi/validators/length.rb
require 'haveapi/validator'

module HaveAPI
  # Checks the length of a string. It does not have a short form.
  #
  # Full form:
  #   string :param, length: {
  #     min: 3,
  #     max: 10
  #     message: 'the error message'
  #   }
  #
  #   string :param, length: {
  #     equals: 8
  #   }
  class Validators::Length < Validator
    name :length
    takes :length

    def setup
      @min = take(:min)
      @max = take(:max)
      @equals = take(:equals)

      if (@min || @max) && @equals
        raise 'cannot mix min/max with equals'

      elsif !@min && !@max && !@equals
        raise 'must use either min, max or equals'
      end

      msg = if @equals
              HaveAPI.message('haveapi.validators.length.equals', equals: @equals)

            elsif @min && !@max
              HaveAPI.message('haveapi.validators.length.min', min: @min)

            elsif !@min && @max
              HaveAPI.message('haveapi.validators.length.max', max: @max)

            else
              HaveAPI.message('haveapi.validators.length.range', min: @min, max: @max)
            end

      @message = take(:message, msg)
    end

    def describe
      ret = {
        message: HaveAPI.localize(@message)
      }

      if @equals
        ret[:equals] = @equals

      else
        ret[:min] = @min if @min
        ret[:max] = @max if @max
      end

      ret
    end

    def valid?(v)
      return false unless v.respond_to?(:length)

      len = v.length

      return len == @equals if @equals
      return len >= @min if @min && !@max
      return len <= @max if !@min && @max

      len.between?(@min, @max)
    end
  end
end
